lib/jerboa/client/worker.ex

```diff
@@ -4,70 +4,166 @@ defmodule Jerboa.Client.Worker do
   use GenServer
 
   alias :gen_udp, as: UDP
-  alias Jerboa.Params
-  alias Jerboa.Format.Body.Attribute.XORMappedAddress
+  alias Jerboa.Client
+  alias Jerboa.Client.Protocol
+  alias Jerboa.Client.Protocol.Transaction
 
-  defstruct [:server, :socket]
+  defstruct [:server, :socket, :mapped_address, :username, :secret,
+             :realm, :nonce, :relayed_address, :lifetime, :lifetime_timer_ref,
+             transaction: %Transaction{}]
+
+  @type socket :: UDP.socket
+  @type state :: %__MODULE__{
+    server: Client.address,
+    socket: socket,
+    transaction: Transaction.t,
+    mapped_address: Client.address,
+    username: String.t,
+    secret: String.t,
+    realm: String.t,
+    nonce: String.t,
+    relayed_address: Client.address,
+    lifetime: non_neg_integer,
+    lifetime_timer_ref: reference
+  }
 
   @system_allocated_port 0
 
-  def start_link(x) do
-    GenServer.start_link(__MODULE__, x)
+  @spec start_link(Client.start_opts) :: GenServer.on_start
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
   end
 
-  def init(address: a, port: p) do
+  ## GenServer callbacks
+
+  def init(opts) do
     false = Process.flag(:trap_exit, true)
     {:ok, socket} = UDP.open(@system_allocated_port, [:binary, active: false])
-    {:ok,
-     %__MODULE__{server: {a, p}, socket: socket}}
+    state = Protocol.init_state(opts, socket)
+    {:ok, state}
   end
 
   def handle_call(:bind, _, state) do
-    msg = Jerboa.Format.encode(Jerboa.Params.put_class(binding_(), :request))
-    response = call(socket(state), server(state), msg)
-    {:ok, params} = Jerboa.Format.decode(response)
-    {:reply, reflexive_candidate(params), state}
+    {result, new_state} =
+      state
+      |> Protocol.bind_req()
+      |> send_req()
+      |> Protocol.eval_bind_resp()
+    {:reply, result, new_state}
+  end
+  def handle_call(:allocate, _, state) do
```

Comment:
Unfortunately no, I can't solve this using `with`, because I won't be able to return `new_state` from one of the else clauses. To be honest, I have no idea how to refactor it at the moment.